### PR TITLE
[nhk] Add support for NHK for School

### PR DESCRIPTION
**Pull request by Lesmiscore**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This adds support for [NHK for School](https://www.nhk.or.jp/school/). Likely not geo-restricted.

Closes ytdl-org/youtube-dl#29115

### CODE REVIEW

File: yt_dlp/extractor/nhk.py [hunk 1]
```diff
@@ -176,3 +183,203 @@ def _real_extract(self, url):
...
+                'title': 'scene 09 『シ』と『ツ』がちがうと…'
+            }, {
+                'start_time': 524.09,
+                'end_time': 555.054,
+                'title': 'scene 10 “ことばドリル”にちょうせん！(2)'
+            }, {
+                'start_time': 555.054,
+                'end_time': 599.999,
+                'title': 'scene 11 うたっておぼえる漢字ドリル'
+            }]
```
Comment (pukkandan) [inline yt_dlp/extractor/nhk.py:249]: needed?
```suggestion
            'chapters': 'count:12'
```

File: yt_dlp/extractor/nhk.py [hunk 2]
```diff
@@ -176,3 +183,203 @@ def _real_extract(self, url):
...
+        'rika', 'syakai', 'kokugo',
+        'sansuu', 'seikatsu', 'doutoku',
+        'ongaku', 'taiiku', 'zukou',
+        'gijutsu', 'katei', 'sougou',
+        'eigo', 'tokkatsu',
+        'tokushi', 'sonota',
+    )
+    _VALID_URL = r'https?://www\.nhk\.or\.jp/school/(?P<id>%s)' % (
+        '|'.join(re.escape(s) for s in KNOWN_SUBJECTS)
+    )
```
Comment (pukkandan) [inline yt_dlp/extractor/nhk.py:315]:
```suggestion
    _VALID_URL = r'https?://www\.nhk\.or\.jp/school/(?P<id>%s)' % '|'.join(re.escape(s) for s in KNOWN_SUBJECTS)
```
etc

File: yt_dlp/extractor/nhk.py [hunk 3]
```diff
@@ -176,3 +183,203 @@ def _real_extract(self, url):
...
+
+    _TESTS = [{
+        'url': 'https://www.nhk.or.jp/school/sougou/',
+        'info_dict': {
+            'id': 'sougou',
+            'title': '総合的な学習の時間',
+        },
+        'playlist_mincount': 16,  # as of 2021/06/20
```
Comment (pukkandan) [inline yt_dlp/extractor/nhk.py:323]:
```suggestion
        'playlist_mincount': 16,
```
etc

File: yt_dlp/extractor/nhk.py [hunk 4]
```diff
@@ -176,3 +183,203 @@ def _real_extract(self, url):
...
+        'info_dict': {
+            'id': 'rika',
+            'title': '理科',
+        },
+        'playlist_mincount': 15,  # as of 2021/06/25
+    }]
+
+    @classmethod
+    def suitable(cls, url):
+        return super(NhkForSchoolSubjectIE, cls).suitable(url) and not NhkForSchoolProgramListIE.suitable(url)
```
Comment (pukkandan) [inline yt_dlp/extractor/nhk.py:335]: isnt it easier to just do this with regex?
```py
r'https?://www\.nhk\.or\.jp/school/(?P<id>%s)(?:$|[#?])'
```

File: yt_dlp/extractor/nhk.py [hunk 5]
```diff
@@ -176,3 +183,203 @@ def _real_extract(self, url):
...
+
+    def _real_extract(self, url):
+        subject_id = self._match_id(url)
+        url = f'https://www.nhk.or.jp/school/{subject_id}/'
```
Comment (pukkandan) [inline yt_dlp/extractor/nhk.py:339]: why?

File: yt_dlp/extractor/nhk.py [hunk 6]
```diff
@@ -176,3 +183,203 @@ def _real_extract(self, url):
...
+
+        webpage = self._download_webpage(url, subject_id)
+
+        return self.playlist_result(
+            [self.url_result(urljoin(url, g.group(1)))
+             for g in re.finditer(
+                rf'href="((?:https?://www\.nhk\.or\.jp)?/school/{re.escape(subject_id)}/[^/]+/")', webpage)],
+            subject_id,
+            self._html_search_regex(r'(?s)<span\s+class="subjectName">\s*<img\s*[^<]+>\s*([^<]+?)</span>', webpage, 'title', fatal=False))
```
Comment (pukkandan) [inline yt_dlp/extractor/nhk.py:348]: wouldnt `playlist_from_matches` be simpler?

File: yt_dlp/extractor/nhk.py [hunk 7]
```diff
@@ -176,3 +183,203 @@ def _real_extract(self, url):
...
+    def _real_extract(self, url):
+        program_id = self._match_id(url)
+
+        webpage = self._download_webpage(f'https://www.nhk.or.jp/school/{program_id}/', program_id)
+
+        # both have format like "Program Name Here | NHK for School", so we have to strip last part
+        _title = self._og_search_title(webpage, fatal=False) or self._html_extract_title(webpage, fatal=False)
+        title = re.sub(r'\s*\|\s*NHK\s+for\s+School\s*$', '', _title) if _title else None
+        if not title:
+            title = self._html_search_regex(r'<h3>([^<]+?)とは？\s*</h3>', webpage, 'title', fatal=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/nhk.py:373]:
```suggestion
        title = self._og_search_title(webpage, fatal=False) or self._html_extract_title(webpage, fatal=False) or self._html_search_regex(r'<h3>([^<]+?)とは？\s*</h3>', webpage, 'title', fatal=False)
        title = re.sub(r'\s*\|\s*NHK\s+for\s+School\s*$', '', title) if title else None
```
(fix formatting)

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/nhk.py (hunk 1 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/nhk.py:249]: not needed maybe

File: yt_dlp/extractor/nhk.py (hunk 4 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/nhk.py:285]: liked that regex, should work as I intended

File: yt_dlp/extractor/nhk.py (hunk 6 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/nhk.py]: a little, but yes

File: yt_dlp/extractor/nhk.py (hunk 5 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/nhk.py]: not needed actually

**Pull request merged**